tango/step.py

```diff
@@ -359,15 +352,250 @@ def unique_id(self) -> str:
                     )
                 )[:32]
             else:
-                self.unique_id_cache += det_hash(random.getrandbits((58 ** 32).bit_length()))[:32]
+                self.unique_id_cache += det_hash(
+                    _random_for_step_names.getrandbits((58 ** 32).bit_length())
+                )[:32]
 
         return self.unique_id_cache
 
     def __hash__(self):
+        """
+        A step's hash is just its unique ID.
+        """
         return hash(self.unique_id)
 
     def __eq__(self, other):
+        """
+        Determines whether this step is equal to another step. Two steps with the same unique ID are
+        considered identical.
+        """
         if isinstance(other, Step):
             return self.unique_id == other.unique_id
         else:
             return False
+
+    def _replace_steps_with_results(self, o: Any, workspace: "Workspace"):
+        if isinstance(o, Step):
+            return o.result(workspace, self)
+        elif isinstance(o, WithUnresolvedSteps):
+            return o.construct(workspace)
+        elif isinstance(o, (list, tuple, set)):
+            return o.__class__(self._replace_steps_with_results(i, workspace) for i in o)
+        elif isinstance(o, dict):
+            return {
+                key: self._replace_steps_with_results(value, workspace) for key, value in o.items()
+            }
+        else:
+            return o
+
+    def result(
+        self, workspace: Optional["Workspace"] = None, needed_by: Optional["Step"] = None
+    ) -> T:
+        """Returns the result of this step. If the results are cached, it returns those. Otherwise it
+        runs the step and returns the result from there.
+
+        If necessary, this method will first produce the results of all steps it depends on."""
+        if workspace is None:
+            from tango.workspace import default_workspace
+
+            workspace = default_workspace
+
+        if self in workspace.step_cache:
+            if click_logger.isEnabledFor(logging.INFO):
+                message = click.style("\N{check mark} Found output for ", fg="green")
+                message += click.style(f'"{self.name}"', bold=True, fg="green")
+                message += click.style(" in cache", fg="green")
+                if needed_by is None:
+                    message += click.style(" ...", fg="green")
+                else:
+                    message += click.style(f' (needed by "{needed_by.name}") ...', fg="green")
+                click_logger.info(message)
+            return workspace.step_cache[self]
```

Comment:
Then I think we should streamline how the logs are configured. I want it so if you do nothing at all, using Tango as a library, you don't see click logs. But if you use the CLI and you do otherwise nothing, you see click logs by default.

I felt a little uneasy about making these into log messages at all, but I think it has to be that way. We're printing these messages from all parts of the system, so any other solution we come up with, we'd have to plumb it through every function call we have. Or we print it all the time, which breaks the as-a-library case. The early transformers library did this, and they finally, thankfully changed it :-)

Reply:
> Then I think we should streamline how the logs are configured. I want it so if you do nothing at all, using Tango as a library, you don't see click logs. But if you use the CLI and you do otherwise nothing, you see click logs by default.

Agreed. This wasn't an issue before because click messages were only emitted from the Executor, so they could easily be turned off.

Reply:
19e6de2